### ISSUE TITLE
Pin (exclude) a directory

### ISSUE THREAD

**Issue by pureby** [microsoft/winget-cli#2749]
### Description of the new feature / enhancement

_This proposal is an extension of #476 by @denelon. Please excuse me if this is seen as a duplicate._

**Users should be able to exclude (or "pin") not only packages, but also _directories_ from `winget upgrade --all`.**

A user might want to install and update certain applications manually, for example in C:\MyApps. Or he might want to rely on application's own auto-update features. If user pins C:\MyApps in Winget only once, he can simply install applications into that directory and be sure they will be automatically excluded from winget upgrade.

This is especially important when using multiple package managers alongside winget (Scoop, Chocolatey, etc.). A user can instruct his secondary package managers to install apps in their dedicated directory (e.g. C:\scoop or C:\choco). This way, whatever applications user installs using his secondary package manager, they will be automatically excluded from winget, so the user does not have to remember to pin each newly installed package manually using `winget pin <package>`.

Another particularly useful scenario for this feature is when a user wants to exclude all games and their launchers (Steam, Ubisoft Connect, Rockstar, etc.) from auto-updating using Winget. He can simply set launchers to install into C:\Games\Steam, C:\Games\Ubisoft, C:\Games\Rockstar, etc. Then the user only has to pin C:\Games in winget, to prevent unwanted game and/or game launcher updates, automatically also for all games installed in the future.

This would also be very handy for excluding development environments from Winget alltogether, by pinning C:\Dev, a developer could install all his dev tools in that folder, possibly multiple versions of the same tool, and not worry about remembering to pin each particular application or version in Winget.

### Proposed technical implementation details

A command like: `winget pin -directory "C:\MyApps"` should exclude all applications installed into _C:\MyApps_ and all of its subfolders from `winget upgrade --all` command. 

A command like `winget unpin -directory "C:\MyApps"` should unpin that directory, making all the applications installed into _C:\MyApps_ and all of its subfolders upgradable using `winget upgrade -all` command.

A command like `winget list --pinned` should list all pinned directories _and_ all pinned applications (if any) IF those applications have explicitly been pinned using `winget pin <package>`. It should _not_ list applications that have not explicitly been pinned using `winget pin <package>`.

If an application installed in _C:\MyApps_ (or any of it's subdirectories) is detected by `winget list` command, a user should still be able to update this application by explicitly using `winget upgrade <package>`.

`winget pin <package>` and `winget unpin <package>` should have prevalence over `winget pin -directory "C:\MyApps"` and `winget unpin -directory "C:\MyApps"` respectively. Meaning that if an application has explicitly been pinned or unpinned using `winget [un]pin <package>`, it's pinned state should remain unchanged by pinning or unpinning the directory into which it has been installed.

**Example:**
- User installs Microsoft Visual Studio Code in `C:\MyApps\VSCode` manually (not using Winget).
- User installs Mozilla Firefox in _C:\MyApps\Firefox_ using `winget install firefox --location "C:\MyApps\Firefox"`.
- `winget upgrade --all` _upgrades both VSCode and Firefox._
- User pins Microsoft Visual Studio Code using `winget pin vscode`.
- User pins _C:\MyApps_ using `winget pin -directory "C:\MyApps"`.
- `winget list --pinned` lists:
- - C:\MyApps
- - Microsoft Visual Studio Code
- `winget upgrade --all` _upgrades neither VSCode nor Firefox._
- User unpins _C:\MyApps_ using `winget unpin -directory "C:\MyApps"`.
- `winget list --pinned` lists:
- - Microsoft Visual Studio Code
- `winget upgrade --all` _upgrades only Firefox._

**Comment by denelon** [comment]
I'm not sure how viable this is as we need a way to reason about a directory which isn't necessarily part of the list of things WinGet reasons about. It's possible it could "detect" the registry entries that indicate a known "installed" package is in a particular directory, but it seems orthogonal to "pinning a package". 

This seems more like the "exclude" mechanism that would apply to packages installed in directories. Maybe it's a nice to have kind of thing, but not necessarily "critical" path for pinning.

**Comment by pureby** [comment]
> This seems more like the "exclude" mechanism that would apply to packages installed in directories. Maybe it's a nice to have kind of thing, but not necessarily "critical" path for pinning.

This is exactly what I was thinking. Winget knows the location of packages it is listing or upgrading. The list of pinned packages and directories can be included in `settings.json`. When `list` or `upgrade` command is initiatied, Winget can simply check the settings file and either perform or not perform an action on certain packages by matching their package name or containing directory against the list in settings.json.

A simple _IF MATCH do nothing ELSE upgrade_ statement will do the trick.

**Comment by masterflitzer** [comment]
after reading the whole proposal i have to say i really like this

just one thing, shouldn't it be `winget [un]pin --directory ` (two dashes before directory)?

**Comment by pureby** [comment]
> after reading the whole proposal i have to say i really like this
> just one thing, shouldn't it be `winget [un]pin --directory` (two dashes before directory)?
You are right, two hyphens make more sense in winget. I updated the OP. Thanks for the feedback.

**Comment by Viexi** [comment]
Any chance this can be implemented in some fashion? I want to run "winget upgrade all" but there are some packages listed that simply must not be upgraded for compatibility purposes

**Comment by denelon** [comment]
* #476 is in progress.
